src/AI/WallyAI.java

```diff
@@ -0,0 +1,788 @@
+package AI;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import CommandingOfficers.Commander;
+import CommandingOfficers.CommanderAbility;
+import Engine.GameAction;
+import Engine.GameAction.ActionType;
+import Engine.GameActionSet;
+import Engine.Path;
+import Engine.Utils;
+import Engine.XYCoord;
+import Engine.Combat.CombatEngine;
+import Terrain.GameMap;
+import Terrain.Location;
+import Terrain.TerrainType;
+import Units.Unit;
+import Units.UnitModel;
+import Units.UnitModel.UnitEnum;
+import Units.Weapons.Weapon;
+import Units.Weapons.WeaponModel;
+
+/**
+ *  Wally makes(?) and breaks walls.
+ */
+public class WallyAI implements AIController
+{
+  Queue<GameAction> actions = new ArrayDeque<GameAction>();
+  // sort our units by expense first, we want them to hit first
+  Queue<Unit> unitQueue = new PriorityQueue<Unit>(11, new AIUtils.UnitCostComparator(false));
+  boolean stateChange;
+
+  private Commander myCo = null;
+
+  // % damage dealable to feel "threatened"
+  private static final int INDIRECT_THREAT_THRESHHOLD = 0;
+  private static final int DIRECT_THREAT_THRESHHOLD = 13;
+  private static final double AGGRO_FUNDS_WEIGHT = 1.5;
+  private static final double RANGE_WEIGHT = 0.07;
+
+  private ArrayList<XYCoord> unownedProperties;
+  private ArrayList<XYCoord> capturingProperties;
+
+  private StringBuffer logger = new StringBuffer();
+  private int turnNum = 0;
+  private int mobileAttackRange = 0;
+
+  public WallyAI(Commander co)
+  {
+    myCo = co;
+    for( UnitModel um : myCo.unitModels )
+    {
+      for( WeaponModel wm : um.weaponModels )
+      {
+        if( wm.canFireAfterMoving )
+          mobileAttackRange = Math.max(mobileAttackRange, wm.maxRange);
```

Comment:
So I can see why you record this and how you use it below, but storing the maximum theoretical attack range across all units as a class-level variable just feels wrong on a very fundamental level.

I mean, it works, and it's your AI, so don't change it if you don't want to, but this smells something awful.

Reply:
Haha.
I guess I could just calculate it as part of the function, but Wally already calculates possibly too much all the time (as you yourself have mentioned the performance impact).

Realistically, we shouldn't be handing out buffs to mobile unit range anyway, so leaving it in the constructor is fine.

Reply:
Oh, man. I just got a _great_ idea for a new CO...

Reply:
I say "shouldn't" because it's nigh-impossible to balance reasonably, not because it's infeasible.

Reply:
Actually; I thought I was just being sarcastic, but how will this play with your proposed battleship rework?

Reply:
We don't have any range-boosting COs, do we?

The B-ship would read as having 5 maximum range, so it would result in Wally calculating more squares than is really needed.

Reply:
> We don't have any range-boosting COs, do we?

None that boost mobile range, no. Who knows how long it'll take to fix that though?

Reply:
> None that boost mobile range, no. Who knows how long it'll take to fix that though?

If you don't count the test branch, we don't have any that boost range at all.
Probably less than an hour.